### DISCUSSION TITLE
How can I extrack the Tempo from a Track chunk?

### DISCUSSION THREAD

**Discussion by fbarrera** [melanchall/drywetmidi#201] [Q&A] [answered]
I'm creating a rhythm game for Unity, and I'd like to extract some information from the MidiEvents that have been triggered. If I loop through the track chunks in the midi file, I can find the midiEvents that are being triggered.

The thing is, I actually find the SetTempo event, and if I print it with ToString() method, the following thing get in the output:

`Set Tempo (468750)`

Question is: how do I get this number? Isn't there any function like GetTempo to extract it from the chunk? Or is there a way to loop through the midi events to extract the data?

There is actually no pointer in the documentation to know how to extract the information from the TempoMap.

Thanks!

**Comment by melanchall** [comment, marked as answer]
First of all, with

```csharp
var tempoMap = midiFile.GetTempoMap();
```

you can get _tempo map_ - an object containing information about all tempo and time signature changes (please read the article in the documentation - https://melanchall.github.io/drywetmidi/articles/high-level-managing/Tempo-map.html).

Then you can get tempo at any point in the time:

```csharp
var startTempo = tempoMap.GetTempoAtTime((MidiTimeSpan)0);
```

or you can get all tempo changes:

```csharp
var tempoChanges = tempoMap.GetTempoChanges();
```

**Reply by LeoMassaFM** [reply]
I know I'm necro-ing this thread, but printing the startTempo variable prints this: "495868 μs/qnote" (in my example, the tempo of the MIDI track at 0 seconds is 121 BPM). How do I make it print only the number part? Thanks.

**Reply by melanchall** [reply]
@LeoMassaFM Just use `BeatsPerMinute` property of the `Tempo`:

```csharp
var bpm = startTempo.BeatsPerMinute;
```

Returned value is a `double`. If you want an integer number, use any desired method. For example, you can just round the value:

```csharp
var bpmRounded = (int)Math.Round(bpm);
```

**Reply by LeoMassaFM** [reply]
Regarding rounding BPM, I've come up with a slight variation to your solution: `var bpmRounded = Math.Round(bpm, 2)`
Which rounds the BPM to the nearest number with a precision of 2 decimal places, which accommodates for the needs of my project (and I assume many others' projects). 
Thanks for the help! You're a lifesaver!

**Reply by LeoMassaFM** [reply]
Oh, and one more thing: if my MIDI file has tempo changes, I can get them with `GetTempoChanges()`. Now, how do I print each new tempo in real time with the `void Update()` function? What I'm trying to say is something like this (rough draft):
`public TempoMap tempoMap;`
`public Tempo currentTempo;    //currentTempo starts as the tempo found at 0 seconds into the midi, then changes after passing through a tempo change`
`public Tempo newTempo;    //newTempo is the tempo of the next tempo change found in the midi.`
`public static MidiFile midiFile;    //the midi file itself`
`public string fileName;    //the name of MIDI file`
`public IEnumerable<ValueChange<Tempo>> tempoChanges;`

`void Start()`
`{`
`        midiFile = MidiFile.Read("Assets/" + fileName);`
`        tempoMap = midiFile.GetTempoMap();`
`        currentTempo = tempoMap.GetTempoAtTime((MidiTimeSpan)0);    //gets the first tempo event in the midi`
`        tempoChanges = tempoMap.GetTempoChanges();`
`}`

`void Update()`
`{`
`        newTempo = tempoMap.GetTempoAtTime((MidiTimeSpan)INSERTVALUEHERE);    //INSERTVALUEHERE is supposed to be the time in microseconds until the next tempo change happens`
`        if (currentTempo != newTempo)`
`            currentTempo = newTempo`
`}`

**Reply by melanchall** [reply]
```csharp
newTempo = tempoMap.GetTempoAtTime(new MetricTimeSpan(INSERTVALUEHERE));
```

INSERTVALUEHERE is a number of microseconds.

But in fact I don't understand what do you want.

> Now, how do I print each new tempo in real time with the void Update() function?

`MidiFile.Read` just reads a file immediately. There is no way to subscribe to reading process and do something when reading engine reads a MIDI event.

**Reply by LeoMassaFM** [reply]
What I wanna do is I want the track's tempo to update in game in real time as each tempo change happens.

**Reply by melanchall** [reply]
> What I wanna do is I want the track's tempo to update in game in real time as each tempo change happens.

What do you mean _"as each tempo change happens"_? Happens where? Again, `MidiFile.Read` just reads a MIDI file immediately. The problem is how to calculate `INSERTVALUEHERE`? If so, I don't know your logic, how you start ticking and so on.

**Reply by LeoMassaFM** [reply]
I got the tempo changes with `tempoMap.GetTempoChanges()`, right? Now, what do I do with it? How can I get all the tempos that this GetTempoChanges got?

**Reply by melanchall** [reply]
Hmm, but in your code you have:

```csharp
tempoChanges = tempoMap.GetTempoChanges();
```

So... what does that mean?

> How can I get all the tempos that this GetTempoChanges got?

You've already got them. Maybe your question is how to get tempo from an item of this collection? If yes, what's wrong with this simple code?

```csharp
foreach (var tempoChange in tempoChanges)
{
    var tempo = tempoChange.Value;
    var bpm = tempo.BeatsPerMinute;
}
```

Just standard C# instructions. Of course you can use LINQ on the collection too.
